File: app03_csharp_nft/backend/NftCerts/Storage/PinataIpfsStorageService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using NftCerts.Config;
using NftCerts.Errors;

namespace NftCerts.Storage;

// Real Pinata pinning via pinFileToIPFS / pinJSONToIPFS. Mirrors app01's PinataIpfsStorageService.
public class PinataIpfsStorageService(PinataProperties properties, IHttpClientFactory httpClientFactory)
    : IIpfsStorageService
{
    private const string PinFilePath = "/pinning/pinFileToIPFS";
    private const string PinJsonPath = "/pinning/pinJSONToIPFS";

    public PinResult PinFile(byte[] content, string filename, string contentType)
    {
        using var form = new MultipartFormDataContent();
        var fileContent = new ByteArrayContent(content);
        fileContent.Headers.ContentType =
            new MediaTypeHeaderValue(contentType.Length > 0 ? contentType : "application/octet-stream");
        form.Add(fileContent, "file", filename);
        return Send(PinFilePath, form, "file");
    }

    public PinResult PinJson(JsonElement jsonPayload, string name)
    {
        string wrapper = JsonSerializer.Serialize(new
        {
            pinataMetadata = new { name },
            pinataContent = jsonPayload,
        });
        var body = new StringContent(wrapper, Encoding.UTF8, "application/json");
        return Send(PinJsonPath, body, "JSON");
    }

    private PinResult Send(string path, HttpContent content, string what)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, properties.BaseUrl.TrimEnd('/') + path)
        {
            Content = content,
        };
        if (properties.HasJwt)
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", properties.Jwt);
        }
        else if (properties.HasApiKeyPair)
        {
            request.Headers.Add("pinata_api_key", properties.ApiKey);
            request.Headers.Add("pinata_secret_api_key", properties.ApiSecret);
        }
        else
        {
            throw new IpfsPinningException(
                "No Pinata credentials configured (PINATA_JWT or PINATA_API_KEY/PINATA_API_SECRET)");
        }

        HttpResponseMessage response;
        string responseBody;
        try
        {
            var client = httpClientFactory.CreateClient("pinata");
            response = client.Send(request, HttpCompletionOption.ResponseContentRead);
            responseBody = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        }
        catch (Exception exception) when (exception is HttpRequestException or TaskCanceledException)
        {
            throw new IpfsPinningException($"Failed to pin {what} to IPFS via Pinata: request failed");
        }

        if (!response.IsSuccessStatusCode)
        {
            throw new IpfsPinningException(
                $"Failed to pin {what} to IPFS via Pinata: HTTP {(int)response.StatusCode}");
        }
        return PinResult.Of(ExtractCid(responseBody));
    }

    internal static string ExtractCid(string responseBody)
    {
        JsonElement root;
        try
        {
            root = JsonDocument.Parse(responseBody).RootElement;
        }
        catch (JsonException)
        {
            throw new IpfsPinningException($"Failed to parse Pinata response: {responseBody}");
        }
        if (!root.TryGetProperty("IpfsHash", out var hash) || hash.GetString() is not { Length: > 0 } cid)
        {
            throw new IpfsPinningException($"Pinata response did not contain an IpfsHash: {responseBody}");
        }
        return cid;
    }
}

File: app02_cpp_nft/backend/src/storage/PinataIpfsStorageService.cpp
#include "storage/PinataIpfsStorageService.h"

#include "error/Exceptions.h"
#include "net/SharedEventLoop.h"
#include "util/Uuid.h"

#include <drogon/HttpClient.h>

#include <sstream>

namespace nftcerts::storage {

namespace {

constexpr const char* kPinFilePath = "/pinning/pinFileToIPFS";
constexpr const char* kPinJsonPath = "/pinning/pinJSONToIPFS";

}  // namespace

PinataIpfsStorageService::PinataIpfsStorageService(config::PinataProperties pinataProperties)
    : pinataProperties_(std::move(pinataProperties)) {}

PinResult PinataIpfsStorageService::pinFile(const std::vector<unsigned char>& content, const std::string& filename,
                                             const std::string& contentType) {
    std::string boundary = "----NftCertsBoundary" + util::randomUuid();

    std::string body;
    body += "--" + boundary + "\r\n";
    body += "Content-Disposition: form-data; name=\"file\"; filename=\"" + filename + "\"\r\n";
    body += "Content-Type: " + (contentType.empty() ? std::string("application/octet-stream") : contentType) +
            "\r\n\r\n";
    body.append(reinterpret_cast<const char*>(content.data()), content.size());
    body += "\r\n--" + boundary + "--\r\n";

    auto client = drogon::HttpClient::newHttpClient(pinataProperties_.baseUrl, nftcerts::net::sharedClientLoop());
    auto req = drogon::HttpRequest::newHttpRequest();
    req->setMethod(drogon::Post);
    req->setPath(kPinFilePath);
    req->addHeader("Content-Type", "multipart/form-data; boundary=" + boundary);

    if (pinataProperties_.hasJwt()) {
        req->addHeader("Authorization", "Bearer " + pinataProperties_.jwt);
    } else if (pinataProperties_.hasApiKeyPair()) {
        req->addHeader("pinata_api_key", pinataProperties_.apiKey);
        req->addHeader("pinata_secret_api_key", pinataProperties_.apiSecret);
    } else {
        throw error::IpfsPinningException(
            "No Pinata credentials configured (PINATA_JWT or PINATA_API_KEY/PINATA_API_SECRET)");
    }
    req->setBody(body);

    auto [result, response] = client->sendRequest(req, 30.0);
    if (result != drogon::ReqResult::Ok || !response) {
        throw error::IpfsPinningException("Failed to pin file to IPFS via Pinata: request failed");
    }
    if (response->getStatusCode() < 200 || response->getStatusCode() >= 300) {
        throw error::IpfsPinningException("Failed to pin file to IPFS via Pinata: HTTP " +
                                           std::to_string(response->getStatusCode()));
    }

    return PinResult::of(extractCid(std::string(response->getBody())));
}

PinResult PinataIpfsStorageService::pinJson(const Json::Value& jsonPayload, const std::string& name) {
    Json::Value wrapper;
    Json::Value metadata;
    metadata["name"] = name;
    wrapper["pinataMetadata"] = metadata;
    wrapper["pinataContent"] = jsonPayload;

    auto client = drogon::HttpClient::newHttpClient(pinataProperties_.baseUrl, nftcerts::net::sharedClientLoop());
    auto req = drogon::HttpRequest::newHttpJsonRequest(wrapper);
    req->setMethod(drogon::Post);
    req->setPath(kPinJsonPath);

    if (pinataProperties_.hasJwt()) {
        req->addHeader("Authorization", "Bearer " + pinataProperties_.jwt);
    } else if (pinataProperties_.hasApiKeyPair()) {
        req->addHeader("pinata_api_key", pinataProperties_.apiKey);
        req->addHeader("pinata_secret_api_key", pinataProperties_.apiSecret);
    } else {
        throw error::IpfsPinningException(
            "No Pinata credentials configured (PINATA_JWT or PINATA_API_KEY/PINATA_API_SECRET)");
    }

    auto [result, response] = client->sendRequest(req, 30.0);
    if (result != drogon::ReqResult::Ok || !response) {
        throw error::IpfsPinningException("Failed to pin JSON to IPFS via Pinata: request failed");
    }
    if (response->getStatusCode() < 200 || response->getStatusCode() >= 300) {
        throw error::IpfsPinningException("Failed to pin JSON to IPFS via Pinata: HTTP " +
                                           std::to_string(response->getStatusCode()));
    }

    return PinResult::of(extractCid(std::string(response->getBody())));
}

std::string PinataIpfsStorageService::extractCid(const std::string& responseBody) {
    Json::CharReaderBuilder builder;
    Json::Value root;
    std::string errs;
    std::istringstream stream(responseBody);
    if (!Json::parseFromStream(builder, stream, &root, &errs)) {
        throw error::IpfsPinningException("Failed to parse Pinata response: " + responseBody);
    }
    if (!root.isMember("IpfsHash") || root["IpfsHash"].asString().empty()) {
        throw error::IpfsPinningException("Pinata response did not contain an IpfsHash: " + responseBody);
    }
    return root["IpfsHash"].asString();
}

}  // namespace nftcerts::storage
